x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expressions.java

```diff
@@ -127,22 +136,48 @@ public static Pipe pipe(Expression e) {
+    }
+
+    public static TypeResolution typeMustBeInteger(Expression e, String functionName, FunctionArgument fArg) {
+        return e.dataType().isNumeric() || DataTypes.isNull(e.dataType()) ?
+            TypeResolution.TYPE_RESOLVED :
+            new TypeResolution(incorrectTypeErrorMessage(e, functionName, fArg, DataTypes.INTEGER));
+    }
+
+    public static TypeResolution typeMustBeNumeric(Expression e, String functionName, FunctionArgument fArg) {
+        return e.dataType().isNumeric() || DataTypes.isNull(e.dataType()) ?
+            TypeResolution.TYPE_RESOLVED :
+            new TypeResolution(incorrectTypeErrorMessage(e, functionName, fArg, DataTypes.NUMERIC));
     }
 
-    public static TypeResolution typeMustBeNumeric(Expression e) {
-        return e.dataType().isNumeric() ? TypeResolution.TYPE_RESOLVED : new TypeResolution(incorrectTypeErrorMessage(e, "numeric"));
+    public static TypeResolution typeMustBeString(Expression e, String functionName, FunctionArgument fArg) {
+        return e.dataType().isString() || DataTypes.isNull(e.dataType()) ?
+            TypeResolution.TYPE_RESOLVED :
+            new TypeResolution(incorrectTypeErrorMessage(e, functionName, fArg, DataTypes.STRING));
     }
 
-    public static TypeResolution typeMustBeNumericOrDate(Expression e) {
-        return e.dataType().isNumeric() || e.dataType() == DataType.DATE ?
+    public static TypeResolution typeMustBeDate(Expression e, String functionName, FunctionArgument fArg) {
+        return e.dataType() == DataType.DATE || DataTypes.isNull(e.dataType())?
             TypeResolution.TYPE_RESOLVED :
-                new TypeResolution(incorrectTypeErrorMessage(e, "numeric", "date"));
+            new TypeResolution(incorrectTypeErrorMessage(e, functionName, fArg, DataTypes.DATE));
     }
-    
-    private static String incorrectTypeErrorMessage(Expression e, String...acceptedTypes) {
-        return "Argument required to be " + Strings.arrayToDelimitedString(acceptedTypes, " or ")
-                + " ('" + Expressions.name(e) + "' type is '" + e.dataType().esType + "')";
+
+    public static TypeResolution typeMustBeNumericOrDate(Expression e, String functionName, FunctionArgument fArg) {
+        return e.dataType().isNumeric() || e.dataType() == DataType.DATE || DataTypes.isNull(e.dataType())?
+            TypeResolution.TYPE_RESOLVED :
+            new TypeResolution(incorrectTypeErrorMessage(e, functionName, fArg, DataTypes.NUMERIC, DataTypes.DATE));
+    }
+
+    private static String incorrectTypeErrorMessage(Expression e,
+                                                    String functionName,
+                                                    FunctionArgument fArg,
+                                                    String... acceptedTypes) {
+        return "'" + functionName + "' requires " + (fArg == null || fArg == FunctionArgument.one_arg ? "" : fArg + " ") +
```

Comment:
The pattern for "quoting" things in ES is to use square brackets - [AVG] requires...`